### ISSUE TITLE
Pyright incorrectly reports that a match on elements of a `UnionType` is not exhaustive

### ISSUE THREAD

**Issue by bradleyharden** [microsoft/pyright#7529]
**Describe the bug**
Pyright incorrectly reports that a match on elements of a `UnionType` is not exhaustive.

**Code or Screenshots**
An example (assumed to be `example.py`) is shown below. I don't how to disambiguate `One` and `Two` as types rather than patterns other than by using the import trick shown. Do you have any suggestion there? Either way, the code runs successfully and prints the expected results, so this seems like a bug to me.

```python
class One:
    pass

class Two:
    pass

type Number = One | Two

def test(number: type[Number]):
    from . import example

    match number:
        case example.One:
            print("one")
        case example.Two:
            print("two")

test(One)
test(Two)
```

Here is the error I get:

```
Cases within match statement do not exhaustively handle all values
  Unhandled type: "type[One] | type[Two]"
  If exhaustive handling is not intended, add "case _: pass"Pylance[reportMatchNotExhaustive](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportMatchNotExhaustive)
(parameter) number: type[One] | type[Two]
```

**VS Code extension or command-line**

I'm running pyright 1.1.354 in VSCode, as part of Pylance version 2024.3.100

**Comment by erictraut** [comment]
I don't understand from your example why you have an import statement in the function `test`. Is `example.One` the same as the class `One` defined earlier in your code sample?

Keep in mind that `type[Number]` is the same as `type[One] | type[Two]`, which represents all class objects that are `One` or `Two` _or subclasses thereof_. I'll also note that the `One` and `Two` classes are not marked `@final`, so it's possible to define subclasses of them.

**Comment by bradleyharden** [comment]
> I don't understand from your example why you have an import statement in the function `test`. Is `example.One` the same as the class `One` defined earlier in your code sample?

Yes, they are the same. I use the import trick, because

```python
match number:
    case One:
        ...
```

is interpreted as an irrefutable pattern that assigns `number` to a new variable named `One`. That's why I asked above:

> I don't how to disambiguate `One` and `Two` as types rather than patterns other than by using the import trick shown. Do you have any suggestion there?

I actually [asked about this on the Python forums a few weeks ago](https://discuss.python.org/t/how-can-you-match-on-a-union-of-types/26785/9). I didn't get a response, though. I came up with this import trick on my own.

> Keep in mind that `type[Number]` is the same as `type[One] | type[Two]`, which represents all class objects that are `One` or `Two` _or subclasses thereof_.

Ah, I didn't think of subclasses. Yeah, I guess that would break my notion of exhaustivness.

> I'll also note that the One and Two classes are not marked `@final`, so it's possible to define subclasses of them.

I was not aware of `@final`, but adding it to `One` and `Two` in this example seems to have no effect. Would you expect it to?

**Comment by erictraut** [comment]
Thanks for the clarification.

Yeah, pyright doesn't currently take into consideration `@final` in this case. It's theoretically possible to add such support, but this is a very atypical pattern, so I don't know if it's worth the effort. It's something I'd consider adding in the future if there's sufficient signal from pyright users that it's desired. For comparison, mypy doesn't support this either.

I've changed this from a bug report to an enhancement request, and I'm going to reject it for now because I don't have any plans currently to add this. If you would like to see this added in the future, feel free to upvote this topic, and I will reopen it in the future if there's sufficient interest.

**Comment by bradleyharden** [comment]
Is there any other way to get exhaustiveness checking over a set of types? Could I reformulate the `match` statement in some way? I'm guessing the answer is no, but I thought I would ask.

Also, just out of curiosity, if pyright were to respect `@final` for this particular case, would it do so only if you had a literal `@final` above the class definition? What if a metaclass were to set `__final__ = True`? I assume the latter would only work at runtime and would not be respected by a static type checker. Is that accurate?

**Comment by erictraut** [comment]
The `match` statement isn't well-suited for exhaustively checking for types (class objects). Its class patterns are implemented with an `isinstance` call rather than an `issubclass`, and you'd normally use the latter for checking types.

One way you can do this is to switch from pattern matching to a more traditional if/elif tree.

```python
def test(number: type[Number]):
    if issubclass(number, One):
        print("one")
    elif issubclass(number, Two):
        print("two")
    else:
        assert_never(number)
```

Your assumption about `@final` and `__final__` is correct. The [typing spec ](https://typing.readthedocs.io/en/latest/spec/qualifiers.html#final)(which is what type checkers are based on) discusses the intended semantics of `@final`, so that's what type checkers honor. `__final__` is a runtime implementation detail, not something that type checkers would honor.

**Comment by bradleyharden** [comment]
Yeah, that's similar to what I had previously, which was basically

```python
if number is One:
    print("one")
elif number is Two:
    print("two")
# ...
```

That worked for my particular use case, where I know there are no subclasses. It's just a bit repetitive, as `number` has to be repeated for each line.

But the real problem is the lack of exhaustiveness checking, which is useful because I have cases with 25+ different types. In fact, while preparing this issue, I found an instance in my real code where I had missed handling a type, so it's not a purely hypothetical concern.

**Comment by erictraut** [comment]
> But the real problem is the lack of exhaustiveness checking

That's why I added the `asert_never` call in the final `else` block. With this call in place, the type checker will statically enforce exhaustiveness.

**Comment by bradleyharden** [comment]
Ah, ok, I didn't think that all the way through. I'm not used to thinking about exhaustiveness in terms of reachability, although the latter is clearly much more general and powerful. I'll have to keep that in mind. Thanks for the tip.

**Comment by erictraut** [comment]
> I'm not used to thinking about exhaustiveness in terms of reachability

This isn't technically using reachability. It's using [type narrowing](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=type-narrowing). Exhaustion is proven by determining If the type is narrowed to `Never`. Type checkers use this same type narrowing technique for proving exhaustion in `match` statements.
